client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -200,8 +201,12 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
+        synchronizer.updateState((state, updates) -> {
```

Comment:
What if updatingConfig is already set to true?

What happens in the following race:
resetReadeGroup(a) - start
resetReadeGroup(b) - start
resetReadeGroup(b) - complete
resetReadeGroup(a) - ???

There is no lock between here and the controller. So an order needs to be defined.

Reply:
that should not be a problem because the call to controller should include the generation and hence idempotent. So if RG(b) has already completed the call, RG(a) will simply repeat it with the generation that has already been included in controller.

Reply:
There is a check now to make sure that we only update `ReaderGroupState` once we have confirmed from the controller that the update was made.

Reply:
Initial config is version 0
resetReadeGroup(a) - Sets flag to 'true'
a is assigned version 1 by the controller.
resetReadeGroup(b)
b is assigned version 2 by the controller.
a unsets the flag and updates the readergroupstate.
b crashes.

Now the controller thinks the version is 2 and the config is b. But the group says A and nobody considers the update ongoing.

Reply:
good point. 
i had originally suggested we manage the RG generation number changes on RG. that takes care of the above issue as the reset workflow is started on rg and terminated on rg and controller communication becomes an intermediate step. 

anyway reiterating the logic here:
updateRg: 
```
// if there are ongoing reset, finish it first before starting the new reset
while (true) {
  if rgstate.updating
    update controller with rgGen and config.
    conditionally reset RG state.  
    continue;
  else 
    newGen = rgGen + 1
    contidionally update -> StartRGReset(newGen, newConfig, updating)
    update controller
    conditionally reset RG state (resets readers and segments and changes updating flag)
    break;
}
// any conditional update failure should go back to the loop.

```
the difference above with the implemented logic is that we generate the "next generation" on RG and update it in rg state first and then in controller.. and after updating the controller we complete the reset of the RG. this way only one RG reset can be happening and only after it completes will the subsequent reset be attempted. and any partially completed reset will be left in that state until another operation on rg is attempted which should find rg is updating state and bring it to completion first before doing anything. and the state is only truly reset as the last step so rg state is never left in an inconsistent state or we dont reset it multiple times. 

@pbelgundi @anirudhkovuru can you guys express the pseudo code of how the implemented logic is supposed to be correct?

Reply:
https://github.com/pravega/pravega/blob/3497a1129e78ecbea223ed579f631d116036eabd/client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java#L207
controller.updateReadergroup is conditional. it will reject the request from (b) if (a) has already incremented the generation to 1. 
and if rg finds the generation on controller ahead of generation in its local state, it should first complete the previously attempted reset and only then should it even attempt the new reset which is to call the controller again to increment the generation.

Reply:
Here is how resetReaderGroup() API on Client would look:
```resetReaderGroup(RGConfig config) {
	1. ss.fetchUpdates();
	2. if (ss.updatingConfig==false) {
	  // this is a conditional update and so for 2 processes trying to do this update in parallel, only 1 will succeed.
	   2.1 val updateFlagResult = ss.updateState(ss.updatingConfig = true)
	   2.2 if (updateFlagResult == SUCCESS) {
		    set config.generation = ss.getConfig().generation;
		    set config.readerGroupId = ss.getConfig().readerGroupId;
		    val controllerUpdateResult = controller.updateRG(scope, rgName, config)
		    if (controllerUpdateResult == SUCCESS) {
			   ss.updateUnconditionally(config)
		    } 
		    // we set ss.updatingConfig = false in the following cases:
		    // (i) If both update to controller and SS have succeeded
		    // (ii) update to Controller has failed, & we never tried updating SS and directly landed here
		    // (iii) in the catch block for `controller.updateRg()` call, so if it throws an exception, we again set back "ss.updatingConfig = false"
		    ss.updateState(ss.updatingConfig = false)
	     } 
	     else {
		      // we're here because updateFlagResult==FAILED
		      // goto 1 
	         }
	} 
	else {
	   // here updatingConfig==true 
	  // goto 1 and keep retrying this reset in a while loop waiting for (updatingConfig == false) or bail out
	}
}
```

Reply:
Additionally, there would be a separate timer thread that would periodically check if config in SS matches with config on Controller and if it does not update SS with config copy from Controller.

Reply:
Implementing the pseudocode as described above.